SI-6690 Release reference to last dequeued element.

retronym opened the pull request:
Avoids leaks in MutableList and its more better known child,
mutable.Queue, when the last element is dequeued or when we
take the tail of a one element collection.

Refactors copy/pasted code between the two implementations of
tail.

Review by @axel22

retronym committed:
SI-6690 Release reference to last dequeued element.

Avoids leaks in MutableList and its more better known child,
mutable.Queue, when the last element is dequeued or when we
take the tail of a one element collection.

Refactors copy/pasted code between the two implementations of
tail.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1824/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1824/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1112/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1112/

axel22 commented on src/library/scala/collection/mutable/MutableList.scala:
```diff
@@ -56,12 +56,16 @@ extends AbstractSeq[A]
-    tl.last0 = last0
     tl.len = len - 1
-    tl
+    tl.last0 = if (tl.len == 0) new LinkedList[A]() else last0
```
Previously we've had an invariant that if `MutableList` is empty, then `last0 == first0`. Now this is true when you create the list, but if you add and then remove from it, this invariant is no longer true.

This may not affect any logic right now, but it's somewhat confusing and it may lead to bugs in the future.
I think it would be good to make `last0` point to a fresh `LinkedList`, `null` or some sentinel singleton right from the start.

retronym replied:
Would it be enough to go for `tl.last0 if (tl.len == 0) tl.first0 else last0`? Or could you elaborate with a patch?

axel22 replied:
I'd either change `var last0: LinkedList[A] = first0` to `var last0: LinkedList[A] = new LinkedList`,
or, even better, change `tailImpl` to:

```
tl.len = len - 1
if (tl.len == 0) {
  tl.first0 = new LinkedList
  tl.last0 = first0 // so that we get a MutableList in the same state as after initialization
} else {
  tl.first0 = first0.tail
  tl.last0 = last0
}
```

axel22 replied:
Or, even better, I think this would avoid leaks as well - change line 68 to:

```
tl.last0 = if (tl.len == 0) tl.first0 else last0
```

EDIT: Now I see that's exactly what you proposed - yes, I think that would be much better.

retronym committed:
Share the empty LinkedList between first0/last0.

This is in line with the existing implementation
of `MutableList#clear`, and in line with an newly
created empty collection.

retronym commented:
@axel22 Yep, I agree. I've implemented that in the latest commit.

axel22 commented:
LGTM!

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1836/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1836/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1124/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1124/
